### ISSUE TITLE
Native UI

### ISSUE THREAD

**Issue by hardliner66** [organix/uFork#56]
I found this project after a deep dive into actor frameworks and I think it's great. I would love to contribute, but I'm not sure my low-level knowledge is enough for all the hard work. But I do have experience with Rust and was thinking, it might be nice to have a native UI as well. Some of the UI frameworks even support WASM, so the same UI code could be used for native and web. I could de-couple the code a bit from WASM itself, so that it runs in both environments.

But before I invest the time, I thought I'd ask first if something like that is even wanted.

**Comment by dalnefre** [comment]
Contributions are certainly welcome. Since there's not much of a published roadmap, it can be challenging to find a good way to participate while staying within the broad scope of the project. I'd be happy to discuss the possibilities with you.

The Rust codebase is, as you've probably noticed, `no-std`. The intent is to have as few dependencies as possible. Besides the primary WASM target, I'm hoping that the VM can be easily compiled for other existing processors such as RiscV or ARM (e.g.: Raspberry Pi) and microcontrollers such as the ESP32. In any case, there is generally no "host" operating system, but rather uFork is running bare-metal, possibly as the foundation for a capability-secure O/S. Any graphics capability would be exposed as a uFork device, with an actor-message interface. I'm not sure if this is what you meant by "native UI".

Another area that could use some development would be to integrate with WASI and implement an abstract host-O/S device. This could provide a generic bridge to systems services when uFork/WASM **_is_** running in a hosted environment. Given the object-capability design of WASI, this might be a good fit. Just something to consider.

LMK what you think. You can find my address on my blog http://www.dalnefre.com/wp/about/

**Comment by hardliner66** [comment]
So, my idea was to keep the `no-std` requirement, but also make it runable as binary on existing systems and port the debug UI to a separate version that can run in browser and natively on linux/windows/mac. This way the VM would be decoupled from the debug UI.

Another thing I was thinking of, was to port some of the utilities over to rust, so they can be re-used from other projects and/or natively on the os as well. Like the assembler or other tools/libs that are currently written in js.

I think having stuff like this in rust would open up the possibility to integrate the VM with a other projects, like Philipp Oppermanns [blog_os](https://os.phil-opp.com/) and provide native toolchains to compile/assemble code for the VM.

> I'm hoping that the VM can be easily compiled for other existing processors such as RiscV or ARM (e.g.: Raspberry Pi) and microcontrollers such as the ESP32.
I have a bit of experience working with armv7 devices, so I can try to make sure it works for that. I also recently gotten some FPGAs just to experiment with stuff like this.

I think having the VM as a library would make it easier to embed it into other applications, including kernels, while still being wasm compatible for web use.

> integrate with WASI and implement an abstract host-O/S device
I think that would tie in with the library approach. This way it could just pass the needed/wanted implementations on initialization, making them available as you go. I'd have to look into WASI itself tho to know what exactly would be needed to make it compatible. I know what WASI is in general, but no in-depth knowledge.

**Comment by dalnefre** [comment]
I would definitely be interest in seeing where you want to go with this. It's not clear to me how all the pieces would fit together, so perhaps a demonstration/proof-of-concept would be a good idea.

I haven't done much toward deploying on the ESP32, but I did find some information on compiling [Rust for the ESP32](https://esp-rs.github.io/book/introduction.html).

Speaking of FPGAs... I'm exploring the possibility of create an FPGA implementation of uFork, which sounds like it might also be interesting to you. I'd like to stick with open-source as much as possible, so my current target platforms are the [Fomu-PVT](https://www.crowdsupply.com/sutajio-kosagi/fomu) and [TinyFPGA-BX](https://www.crowdsupply.com/tinyfpga/tinyfpga-ax-bx).

**Comment by hardliner66** [comment]
I already started a branch and should have a prototype soon.

As for FPGAs:
I got the [go board](https://nandland.com/the-go-board/) from nandland.com which is fully open source AFAIK.

I also got the [tang primer 20k](https://wiki.sipeed.com/hardware/en/tang/tang-primer-20k/primer-20k.html) from sipeed. This one isn't usable with open source tools, but it's also quite cheap for what it provides.

**Comment by dalnefre** [comment]
I anticipate that it will be a tight squeeze to fit a uFork core on a 1k-LUT device. However, if we keep that goal in mind we could end up with the option of instantiating multiple cores on a 5k or 8k device, which could be very cool.

I look forward to tracking your progress. Feel free to reach out via email too.

**Comment by hardliner66** [comment]
Okay, so I have got to a point where the VM starts and can run the fibonacci program. I had to cheat a bit and extract the memory from the old version, because the assembler is still in javascript, but it works.

I have also tested it in the browser, where it kinda works. The execution works fine, but theres still problems with logging and stuff. Sadly, it also doesn't work from my github pages or I could demo it, but that should be somewhat easy to fix.

The changes are at https://github.com/hardliner66/uFork/tree/restructuring and the whole thing can be run with a simple `cargo run`. To test the web version you need trunk, but then you can do `trunk serve crates/debug-ui/index.html` from the repo root and it should start a server.

I also created a wrapper-lib that's similar to the original code, which in theory could be used as a drop-in replacement of the old wasm, but I didn't set it up to build yet. Once that's done, it should be possible to use the same core for the your frontend and mine.

**Comment by hardliner66** [comment]
> I like where you're going with this! The separation of concerns and increase in modularity are most welcome.

Thanks. I think having a good user experience is paramount and rust makes it relatively easy to have something run on desktop, web and bare metal. So having a flexible core opens up many possibilities for others to use and contribute to the project. And I think that's possible while keeping the original guarantee of `no_std`.

I think it would also be a good idea to parametrize the internal memory sizes, otherwise it might be hard to fit the same code on a small micro controller with less memory available.

> make them available from the command-line, probably using node or deno.

I'd probably do them in rust as well. Create libs for parsing and compiling and write frontends for it. Then those tools can be used as universally as the core VM. That means you can just mix and match which parts of the project you want and integrate that into, say, an os.

I'm also thinking about maybe doing my own take on an actor language, but that's a topic for a different time.

> Just a heads-up that I'm at the beginning of a rather substantial systemic overhaul/refactoring

As long as the structure of the code itself stays similar and it's just internals that get changed, this should not be a problem. If the rewrite gets too big, I can still reuse most of the wrapper code and just rewrite the new code again to fit the new model. So that shouldn't be too much of a problem either.

> Longer-term, there's a plan to extract the blob device (including blob memory) from the core

I've already started moving the implementation of those things outside of the core, so that each host can just pass them as lambdas. This might need some reworking in the future, but it's a good step to having the devices not coupled to the core.

> and re-organizing the error codes 

IMO, this is something that could even be rustified a bit, by using proper enums and `Result<T, ErrorCode>` inside the core and mapping it to integer values at the boundary. There's a few other cases like that where better rust solutions exist, that should only be mapped on the boundary. Some of these rustifications can even provide better optimizations, tho all of them should bring better ergonomics and better safety guarantees. (e.g.: not returning a value that's not a defined ErrorCode.

**Comment by hardliner66** [comment]
I decided to take the learnings from my branch and make a smaller commit that keeps the current UI working. Once that is merged, we can work from there to get the other features going. I'm also gonna create a few more issues for other stuff I wanna tackle, to better keep track of the single steps. For everything else I'm gonna write a mail :)
